torch/_inductor/kernel/flex_attention.py

```diff
@@ -342,77 +342,163 @@ def forward_inner(
-        qk = tl.dot(q, k) # TODO: use cuda matmul when q_len <= 2.
-        if not PRESCALE_QK:
-            qk *= SM_SCALE
-        # ~~~~~~~~~~~~~~~~~~~ Apply score modification  ~~~~~~~~~~~~~~~~~~~
+    if not IS_DIVISIBLE:
+        if block_n_end >= 1:
+            # loop over k, v and update accumulator until block_n_end - 1
+            for start_n in range(block_n_start, block_n_end - 1):
+                acc, l_i, m_i = fwd_compute_block_mn(
+                    q, K_block_ptr, V_block_ptr, Q_LEN, KV_LEN,
+                    # accumulated values
+                    acc, l_i, m_i,
+                    # Offsets
+                    off_z, off_h, offs_m, offs_n,
+                    MATMUL_PRECISION, RCP_LN2,
+                    {{gen_argdefs()}},
+                    IS_FULL_BLOCKS,
+                )
+
+                # update pointers
+                offset = get_offset_for_next_block(
+                    start_n, kv_indices, kv_num_blocks,
+                    SPARSE_KV_BLOCK_SIZE, SPARSE_KV_MULTIPLE, BLOCK_N
+                )
+
+                V_block_ptr = tl.advance(V_block_ptr, (offset, 0))
+                K_block_ptr = tl.advance(K_block_ptr, (0, offset))
+
+                offs_n = offs_n + offset
+
+            # Need to mask out part of the lask block for non divisible seqlen.
+            acc, l_i, m_i = fwd_compute_block_mn(
+                q, K_block_ptr, V_block_ptr, Q_LEN, KV_LEN,
+                # accumulated values
+                acc, l_i, m_i,
+                # Offsets
+                off_z, off_h, offs_m, offs_n,
+                MATMUL_PRECISION, RCP_LN2,
+                {{gen_argdefs()}},
+                IS_FULL_BLOCKS, True,
+            )
+
+
+    else:
+        # loop over k, v and update accumulator until block_n_end
+        for start_n in range(block_n_start, block_n_end):
+            acc, l_i, m_i = fwd_compute_block_mn(
+                q, K_block_ptr, V_block_ptr, Q_LEN, KV_LEN,
+                # accumulated values
+                acc, l_i, m_i,
+                # Offsets
+                off_z, off_h, offs_m, offs_n,
+                MATMUL_PRECISION, RCP_LN2,
+                {{gen_argdefs()}},
+                IS_FULL_BLOCKS,
+            )
+
+            # update pointers
+            offset = get_offset_for_next_block(
+                start_n, kv_indices, kv_num_blocks,
+                SPARSE_KV_BLOCK_SIZE, SPARSE_KV_MULTIPLE, BLOCK_N
+            )
+
+            V_block_ptr = tl.advance(V_block_ptr, (offset, 0))
+            K_block_ptr = tl.advance(K_block_ptr, (0, offset))
+
+            offs_n = offs_n + offset
+
+    return acc, l_i, m_i
+
+"""
+
+
+fwd_compute_block_mn = r"""
+@triton.jit
+def fwd_compute_block_mn(
+    q, K_block_ptr, V_block_ptr, Q_LEN, KV_LEN,
+    # accumulated values
+    acc, l_i, m_i,
+    # Offsets
+    off_z, off_h, offs_m, offs_n,
+    MATMUL_PRECISION, RCP_LN2,
+    {{gen_argdefs()}},
+    IS_FULL_BLOCKS, is_last_block=False,
```

Comment:
I'll do a refactor to organize these kernel functions and clean up arguments after this PR.